icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java

```diff
@@ -526,14 +530,127 @@ FormattedListBuilder formatImpl(Collection<?> items, boolean needsFields) {
         builder.append(start, it.next(), 1);
         for (int idx = 2; idx < count - 1; ++idx) {
             builder.append(middle, it.next(), idx);
         }
-        return builder.append(end, it.next(), count - 1);
+        Object last = it.next();
+        return builder.append(patternHandler.getEndPattern((String)last), last, count - 1);
+    }
+
+    // A static hanlder append the items without considering the text inside the it
+    private class StaticHandler implements PatternHandler {
+        StaticHandler(String two, String end) {
+            this.twoPattern = two;
+            this.endPattern = end;
+        }
+
+        public String getTwoPattern(String text) { return this.twoPattern; }
+        public String getEndPattern(String text) { return this.endPattern; }
+
+        private final String twoPattern;
+        private final String endPattern;
+    }
+
+    // A contextual hanlder append the items by as static handler, except delegate to appender to
+    // handle the the two and end pattern.
+    private class ContextualHandler implements PatternHandler {
+        ContextualHandler(Pattern regexp, String thenTwo, String elseTwo, String thenEnd, String elseEnd) {
+            this.regexp = regexp;
+            this.thenTwoPattern = thenTwo;
+            this.elseTwoPattern = elseTwo;
+            this.thenEndPattern = thenEnd;
+            this.elseEndPattern = elseEnd;
+        }
+
+        public String getTwoPattern(String text) {
+            if(this.regexp.matcher(text).matches()) {
+                return this.thenTwoPattern;
+            } else {
+                return this.elseTwoPattern;
+            }
+        }
+
+        public String getEndPattern(String text) {
+            if(this.regexp.matcher(text).matches()) {
+                return this.thenEndPattern;
+            } else {
+                return this.elseEndPattern;
+            }
+        }
+
+        private final Pattern regexp;
+        private final String thenTwoPattern;
+        private final String elseTwoPattern;
+        private final String thenEndPattern;
+        private final String elseEndPattern;
+
+    }
+
+    // Pattern in the ICU Data which might be replaced y by e.
+    private static final String compiledY = compilePattern("{0} y {1}", new StringBuilder());
+
+    // The new pattern to replace y to e
+    private static final String compiledE = compilePattern("{0} e {1}", new StringBuilder());
+
+    // Pattern in the ICU Data which might be replaced o by u.
+    private static final String compiledO = compilePattern("{0} o {1}", new StringBuilder());
+
+    // The new pattern to replace u to o
+    private static final String compiledU = compilePattern("{0} u {1}", new StringBuilder());
+
+    // Condiction to change to e.
+    // Starts with "hi" or "i" but not with "hie" nor "hia"a
+    private static final Pattern changeToE = Pattern.compile("(?i)^((i|hi[^ae]).*|hi)$");
```

Comment:
Optional:
- You don't need the ^ and $ in the patterns because matches() wants to cover the whole string.
- I think you can simplify to `"(?i)(i.*|hi[^ae]*)"`
- even simpler: `Pattern.compile("(i.*|hi[^ae]*)", CASE_INSENSITIVE)`

Reply:
but ... (i.*|hi[^ae]*) will return hioa false, right?

Reply:
You are right, sorry for my bad regex. Your new version looks good.